concordium-consensus/src/Concordium/TransactionVerification.hs

```diff
@@ -0,0 +1,347 @@
+import qualified Data.Serialize as S
+import Lens.Micro.Platform
+
+import qualified Concordium.Types.Transactions as Tx
+import qualified Concordium.ID.AnonymityRevoker as AR
+import qualified Concordium.ID.IdentityProvider as IP
+import qualified Concordium.GlobalState.Types as GSTypes
+import qualified Concordium.GlobalState.BlockState as BS
+import qualified Concordium.Types.Parameters as Params
+import qualified Concordium.Types.Updates as Updates
+import qualified Concordium.Types as Types
+import qualified Concordium.ID.Account as A
+import qualified Concordium.ID.Types as ID
+import qualified Concordium.Crypto.SHA256 as Sha256
+import Concordium.Types.HashableTo (getHash)
+import qualified Concordium.Cost as Cost
+
+import Data.Maybe (isJust)
+import Control.Monad.Except
+import Concordium.Types.Updates (UpdateSequenceNumber)
+
+-- |VerificationResults of transactions.
+-- A verification result can either be 'Ok', 'MaybeOk' or 'NotOk'.
+-- 
+-- This type exists as an intermediate type for decoupling the internal verification process from the outcome of processing a transaction.
+data VerificationResult
+  = Ok OkResult
+  -- ^The transaction successfully passed the verification check.
+  | MaybeOk MaybeOkResult
+  -- ^The transaction did not pass the verification check, but it might
+  -- be valid in a later block. As such if the transaction was received individually
+  -- it must be rejected at once. However if the transaction was received as part of a block,
+  -- then it must be accepted and re-verified in the 'Scheduler'.
+  | NotOk NotOkResult
+  -- ^The transaction is definitely not valid and must be rejected.
+  deriving (Eq, Show, Ord, Generic)
+instance S.Serialize VerificationResult
+
+-- |The transaction was verified successfully and was valid at the point of receiving it.
+-- The OkResult's should contain enough information for the scheduler to verify if it is still valid
+-- or if it needs a new verification before executing.
+--
+-- 'CredentialDeploymentSuccess' as ip's or ar's cannot be removed then nothing needs to be checked before executing
+-- 'ChainUpdateSuccess' it should be checked that the authorization keys matches the current ones before executing
+-- 'NormalTransactionSuccess' it should be checked that the account information can still verify  before executing
+data OkResult
+  = CredentialDeploymentSuccess
+  -- ^The 'CredentialDeployment' passed verification.
+  | ChainUpdateSuccess !Sha256.Hash !Updates.UpdateSequenceNumber
+  -- ^The 'ChainUpdate' passed verification.
+  -- The result contains the hash of the autorization keys.
+  -- It must be checked that this hash corresponds to the current authorization keys before
+  -- executing the transaction.
+  | NormalTransactionSuccess !Sha256.Hash !Types.Nonce
+  -- ^The 'NormalTransaction' passed verification.
+  -- The result contains the hash of the keys and the transaction nonce.
+  deriving (Eq, Show, Ord, Generic)
+instance S.Serialize OkResult
```

Comment:
Why is serialization of this even needed? 

If it is not, remove it. If it is, then it should be spelled out manually so that it is stable under renamings, restructurings, of the type.